### ISSUE TITLE
Parsing for iPhone iTunes Backups

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#977]
This is to track the continuation of the initial work on #517, since that was scheduled for 4.0, but this is not.

**Comment by paulobreim** [comment]
This feature is very important.

**Comment by lfcnassif** [comment]
I agree, like many others in the long queue... If you could help implementing this, your contribution would be welcomed.

**Comment by gfd2020** [comment]
Is this issue still open?
I believe the solution below could help with the issue.

https://github.com/jsharkey13/iphone_backup_decrypt

**Comment by lfcnassif** [comment]
It is, thanks for pointing.

**Comment by gfd2020** [comment]
Hi @lfcnassif .

I found this Java lib ( https://github.com/MaxiHuHe04/iTunes-Backup-Explorer ) to help decode iTunes backups. It's Java-based. I ran some tests, and it's quite easy to extract files using the lib.
However, there are some challenges in adapting it to IPED; at least I don't know how to get started. I'd like some help.

1) I need to identify the plist file (so far, so good).
2) Once the plist is identified, I need to find the Manifest.db file, which is in the same folder as the plist.
3) With these two files saved in a temporary folder, I need to open the database.
4) The next files that appear in this folder or in subdirectories should be redirected to the parser to be queried in the database by name. If successful, they will be decrypted and redirected to a specific parser.

Question:
Is there a similar parser I can use?

**Comment by lfcnassif** [comment]
Hi @gfd2020!

Yes, I've found that library before, it is around for some time and is very popular, seems a good option.

> Is there a similar parser I can use?

You can take a look at any parser that uses the IIPEDSearcher interface, like WhatsAppParser (very complex), TelegramParser, EmuleParser, etc. That interface is used to search for other items in the case and later you can get their properties or content. You may need (or not) to update the MimeTypesProcessingOrder class, because you can just search for items processed before in previous processing queues, so maybe your main item (plist or DB?) may need to be processed in a later queue.

> 4. The next files that appear in this folder or in subdirectories should be redirected to the parser to be queried in the database by name. If successful, they will be decrypted and redirected to a specific parser.

Actually you don't need to redirect those files to the parser, but you will search for them within the parser with interface above.

What I think is challenging here is taking renamed files, in a totally different folder tree, and presenting them to the user in their original paths and names, without processing the same content twice... At least I think it would be the perfect result to display to users... This way, file paths and names could also be searched from future iOS specific parsers. Not sure if it is possible to achieve this without "duplicating" files in their original paths and processing them again. And keeping the original backup structure in the evidence seems important too...

**Comment by gfd2020** [comment]
Thanks @lfcnassif 

> What I think is challenging here is taking renamed files, in a totally different folder tree, and presenting them to the user in their original paths and names, without processing the same content twice... At least I think it would be the perfect result to display to users... This way, file paths and names could also be searched from future iOS specific parsers. Not sure if it is possible to achieve this without "duplicating" files in their original paths and processing them again. And keeping the original backup structure in the evidence seems important too...

I don't know if I understood the difficulty very well but what I thought of doing would be something I did in the OFX parser. Given an encrypted iTunes file, it would have a decrypted derivation, for example:

itunes_backup/00/000a2dc77b04f8feea3f7aff9ea96ca753e2af84>>domain/real_path/real_file_name.jpg

As for the directories, it wouldn't extract them by default, only if the user configured it.

**Comment by lfcnassif** [comment]
Hi @gfd2020!

I think your proposal should work for most cases, thanks! But I should point some features wouldn't work, like navigation through the File Tree, Go to Parent Folder to see siblings, Subitems Panel listing for parent folders, searches for parentId:X on UI and from parsers.

Maybe an alternative solution would be creating a virtual file tree using the original paths and file names, below the plist or the Manifest.db file. The parser could put into each derivated item a metadata to store the backup item path or id from where it was derivated from, so we wouldn't loose this important relationship.

Votes?

**Comment by gfd2020** [comment]
> Hi [@gfd2020](https://github.com/gfd2020)!
> 
> I think your proposal should work for most cases, thanks! But I should point some features wouldn't work, like navigation through the File Tree, Go to Parent Folder to see siblings, Subitems Panel listing for parent folders, searches for parentId:X on UI and from parsers.
> 
> Maybe an alternative solution would be creating a virtual file tree using the original paths and file names, below the plist or the Manifest.db file. The parser could put into each derivated item a metadata to store the backup item path or id from where it was derivated from, so we wouldn't loose this important relationship.
> 
> Votes?

You are correct. I think your suggestion would be the most appropriate.

**Comment by paulobreim** [comment]
I am willing to help with the tests.

**Comment by lfcnassif** [comment]
As a final user, which approach do you prefer @paulobreim ?

**Comment by paulobreim** [comment]
I have an iPhone cloned using Oxygen Forensics, which, when performing the cloning process, recorded the content via iTunes. So I can compare the messages it generates with those we would generate on IPED, reading iTunes.
Oxygen also does voice transcription, but I don't yet know what technology it uses.
So I can use a copy of the IPED for this purpose.

**Comment by gfd2020** [comment]
> I have an iPhone cloned using Oxygen Forensics, which, when performing the cloning process, recorded the content via iTunes. So I can compare the messages it generates with those we would generate on IPED, reading iTunes. Oxygen also does voice transcription, but I don't yet know what technology it uses. So I can use a copy of the IPED for this purpose.

Thank you @paulobreim .
But how would you like file information to be displayed in the interface, in terms of directory and file structure?

**Comment by paulobreim** [comment]
Well, it's like this:
On "joao's" MacBook, there are these WhatsApp chats that were processed correctly by IPED.
They're in a structure like this:
joao/Library/Group Containers/group.net.whatsapp.WhatsApp.shared/ChatStorage.sqlite>>WhatsApp Group - SUPER CARROS ALPHAVILLE 1 - 5511999999999-1585953273_5

I don't know if this was saved by the WhatsApp desktop app. These same chats appear in Oxygen, which retrieved them from João's iPhone.

When I indexed the iPhone using the same iTunes backup generated by Oxygen, it appears that only the WhatsApp chats weren't processed, so if you can maintain the same structure, that would be ideal.

If you want to look at this structure that's already on my server, we can arrange to do that someday.

Paulo

**Comment by celsonaemen** [comment]
Olá, pessoal! Gostaria de contribuir com esta issue e vou começar estudando os pontos de integração com o IPED.

Pretendo analisar o código do [iTunes Backup Explorer](https://github.com/MaxiHuHe04/iTunes-Backup-Explorer), já citado pelo @gfd2020, para entender como aproveitar sua lógica de leitura do Manifest.db e relacionar os arquivos do backup aos seus nomes e caminhos originais.

Como primeiro recorte, quero focar em backups sem criptografia e avaliar a árvore virtual sugerida pelo @lfcnassif, preservando a referência de cada item ao arquivo correspondente no backup. Isso poderia facilitar a navegação e ajudar os parsers que dependem dos caminhos originais a localizar seus arquivos. Também pretendo investigar como evitar o processamento duplicado do conteúdo.

Numa etapa posterior, gostaria de avaliar o suporte a backups criptografados quando a senha estiver disponível, dependendo do que encontrarmos nessa análise inicial.

Existe alguma implementação ou branch em andamento que eu possa consultar para aproveitar o trabalho já realizado e evitar duplicação de esforços? Vou começar pela pesquisa e compartilhar os resultados e uma proposta mais concreta conforme entender os pontos de integração.

**Comment by lfcnassif** [comment]
Hi @celsonaemen! Thank you very much for helping with this support!

As far as I know, we don't have an existing branch with a proposal for this.

**Comment by celsonaemen** [comment]
Hi @lfcnassif, thanks again for the reply.

Before opening a PR, I prepared and validated a small local first step for this issue. It adds a dedicated parser for the already detected `Manifest.db` and produces a searchable catalog with `fileID`, domain, relative path, entry type, flags, metadata blob size, and the total number of entries.

I kept the scope limited for now. This version does not resolve the physical file by `fileID`, build a virtual backup tree, decode the NSKeyedArchiver metadata blob, or handle encrypted backups. My concern was to avoid depending on processing order or duplicating evidence content before the integration approach is clear.

The parser uses IPED's existing temporary SQLite copy flow, falls back to the generic SQLite parser when the expected schema is not available, and has a synthetic `Manifest.db` test. The affected modules compile, and the focused test passes without using real backup or case data.

Would this catalog-only parser be useful as a first PR for review, or would you prefer that the initial proposal also include the `fileID`-to-physical-content association? I have not pushed the branch or opened a PR yet.

**Comment by celsonaemen** [comment]
One hypothetical example of the intended use: an iTunes/Finder backup may store a file under an identifier such as `3a/3a72...`, while the corresponding `Files` row records a logical location such as `HomeDomain/Library/SMS/sms.db`.

With this initial parser, an examiner could search the generated catalog for terms such as `SMS`, `CallHistory`, `DCIM`, or an application domain and see the logical path together with its recorded `fileID`, instead of manually querying and interpreting the `Files` table.

This example is only about catalog navigation. The parser does not open or interpret the referenced payload, does not claim that missing content was deleted, and does not recover data that is absent from the backup.
